waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -82,22 +79,35 @@ data TypeCoercionErrorReason e
     ReasonDictWrongKeyType String e
   deriving (Eq, Show)
 
-getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
-getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
+second :: (b -> d) -> (a, b, c) -> (a, d, c)
+second f (x, y, z) = (x, f y, z)
+
+extractTypeErrorMessagesAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, [String], Ctx)
+extractTypeErrorMessagesAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
   case reason of
     ReasonList e ->
-      first (("For list element:\n" ++) . indent 2) $
-        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+      second ("-> In list element" :) $ extractTypeErrorMessagesAndCtx getUncoercableTypesMsg e
     ReasonDictWrongKeyType key e ->
-      first ((("For dictionary field '" ++ key ++ "':\n") ++) . indent 2) $
-        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
-    ReasonDictNoKey key -> ("Missing required dictionary field '" ++ key ++ "'", ctx)
-    ReasonDictExtraKey key -> ("Unexpected dictionary field '" ++ key ++ "'", ctx)
+      second (("-> In dictionary field '" ++ key ++ "'") :) $ extractTypeErrorMessagesAndCtx getUncoercableTypesMsg e
+    ReasonDictNoKey key -> ("-> Missing required dictionary field '" ++ key ++ "'", [], ctx)
+    ReasonDictExtraKey key -> ("-> Unexpected dictionary field '" ++ key ++ "'", [], ctx)
     ReasonDecl -> uncoercableTypesMsgAndCtx
     ReasonEnum -> uncoercableTypesMsgAndCtx
     ReasonUncoercable -> uncoercableTypesMsgAndCtx
   where
-    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, ctx)
+    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, [], ctx)
+
+joinAdditionalMessages :: [String] -> String
+joinAdditionalMessages [] = ""
+joinAdditionalMessages msgChain = foldr1 appendMsg msgChain
```

Comment:
Thanks for the doc! I'll take a look and decide.

> Since you have this function, I would also move the " -> " here, so it becomes part of "concatenation" logic, not the part of actual error message production logic.

Hm, I thought about this, but decided against it. We prefix all messages with arrows, even the first one (i.e., it's not a part of the concatenator). Still, it's nicer not having to repeat the arrow four times. I think I'll add it here.